interfaces/input_parameters/ring.py

```diff
@@ -182,182 +130,348 @@ class Ring:
             self.Particle.mass,
-            self.ring_circumference, sample_func,
+            self.circumference, sample_func,
             interpolation, start, stop,
             store_turns=store_turns)
 
-        self.n_sections = self.momentum.shape[0]-2
-        self.cycle_time = self.momentum[1]
+        # Getting the cycle time from the interpolation
+        self.cycle_time = momentum_processed[1]
+
+        # The machine turn numbers corresponding to the cycle_time
+        # are kept if the store_turns is enabled
         if store_turns:
             self.parameters_at_turn = self._parameters_at_turn
-            self.use_turns = self.momentum[0].astype(int)
+            self.use_turns = momentum_processed[0].astype(int)
         else:
             self.parameters_at_turn = self._no_parameters_at_turn
-            self.use_turns = self.momentum[0]
+            self.use_turns = momentum_processed[0]
+
         # Updating the number of turns in case it was changed after ramp
         # interpolation
-        self.n_turns = self.momentum.n_turns
-
-        # Derived from momentum
-        self.beta = rt.mom_to_beta(self.momentum[2:], self.Particle.mass)
-        self.gamma = rt.mom_to_gamma(self.momentum[2:], self.Particle.mass)
-        self.energy = rt.momentum_to_energy(self.momentum[2:],
-                                            self.Particle.mass)
-        self.kin_energy = rt.momentum_to_kin_energy(self.momentum[2:],
-                                                    self.Particle.mass)
-        self.t_rev = np.dot(self.ring_length, 1/(self.beta*c))
+        self.n_turns = momentum_processed.n_turns
+
+        # Generating the attributes where to store the synchronous data
+        # and associated parameters
+        program_shape = (self.n_sections, self.n_turns)
+        self.momentum = ring_programs.momentum_program.zeros(program_shape)
+        self.energy = ring_programs.total_energy_program.zeros(program_shape)
+        self.kin_energy = ring_programs.kinetic_energy_program.zeros(
+            program_shape)
+        self.beta = ring_programs._synchronous_data_program.zeros(
+            program_shape)
+        self.gamma = ring_programs._synchronous_data_program.zeros(
+            program_shape)
+
+        # Populating the synchronous data attribute each row corresponding
+        # to a section
+        for index_section in range(self.n_sections):
+            self.momentum[index_section, :] = momentum_processed[2:]
+            self.beta[index_section, :] = rt.mom_to_beta(
+                momentum_processed[2:], self.Particle.mass)
+            self.gamma[index_section, :] = rt.mom_to_gamma(
+                momentum_processed[2:], self.Particle.mass)
+            self.energy[index_section, :] = rt.momentum_to_energy(
+                momentum_processed[2:], self.Particle.mass)
+            self.kin_energy[index_section, :] = rt.momentum_to_kin_energy(
+                momentum_processed[2:], self.Particle.mass)
```

Comment:
This can all be done as
```python
self.momentum = momentum_processed[2:]
self.energy = momentum.to_energy(args, kwargs)
#etc
```
Just needs the `beta` and `gamma` functions adding in the `ring_programs.data_type`

Reply:
Yes that's how it is in the present version, took a long detour to come back to the starting point.

Just figured in the meanwhile that we need to homogenize the definition of 'turns' from datatypes/Ring.